# eda.ipynb file for EGT 309 Project

Credits:

Richie:
- Completed 1.1 to 1.6 (Inspected file 1 to file 6.)
- Completed 2.1, 2.2 data cleaning for file 5.

Xiu Wen:
- Created format for EDA file (sections, titles, markdown positions)
- Completed 1.7 to 1.9 (Inspected file 7 to file 9.)
- Completed 2.3 data cleaning for file 7.

## 1. Data Loading and Initial Inspection

In [1]:
# Import Libraries.
import pandas as pd
import numpy as np
# import matplotlib.pyplot as plt           commented out to avoid import errors on vsc first 
# import seaborn as sns

In [2]:
# Note: We did not read the data_dictionary as it is not a dataset to be used. Viewed it externally on excel instead.

customers_dataset = pd.read_csv('data/olist_customers_dataset.csv')
geolocation_dataset = pd.read_csv('data/olist_geolocation_dataset.csv')
order_items_dataset = pd.read_csv('data/olist_order_items_dataset.csv')
order_payments_dataset = pd.read_csv('data/olist_order_payments_dataset.csv')
order_reviews_dataset = pd.read_csv('data/olist_order_reviews_dataset.csv')
orders_dataset = pd.read_csv('data/olist_orders_dataset.csv')
products_dataset = pd.read_csv('data/olist_products_dataset.csv')
sellers_dataset = pd.read_csv('data/olist_sellers_dataset.csv')
product_category_name_translation_dataset = pd.read_csv('data/product_category_name_translation.csv')

### 1.1 Inspect customers_dataset

In [3]:
customers_dataset

customer_id                customer_unique_id  \
0      06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1      18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2      4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3      b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4      4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
...                                 ...                               ...   
99436  17ddf5dd5d51696bb3d7c6291687be6f  1a29b476fee25c95fbafc67c5ac95cf8   
99437  e7b71a9017aa05c9a7fd292d714858e8  d52a67c98be1cf6a5c84435bd38d095d   
99438  5e28dfe12db7fb50a4b2f691faecea5e  e9f50caf99f032f0bf3c55141f019d99   
99439  56b18e2166679b8a959d72dd06da27f9  73c2643a0a458b49f58cea58833b192e   
99440  274fa6071e5e17fe303b9748641082c8  84732c5050c01db9b23e19ba39899398   

       customer_zip_code_prefix          customer_city customer_state  
0                         14409                 franca             SP  
1                          9790  sao bernardo do campo             SP  
2                          1151              sao paulo             SP  
3                          8775        mogi das cruzes             SP  
4                         13056               campinas             SP  
...                         ...                    ...            ...  
99436                      3937              sao paulo             SP  
99437                      6764        taboao da serra             SP  
99438                     60115              fortaleza             CE  
99439                     92120                 canoas             RS  
99440                      6703                  cotia             SP  

[99441 rows x 5 columns]

In [4]:
customers_dataset.dtypes

customer_id                 object
customer_unique_id          object
customer_zip_code_prefix     int64
customer_city               object
customer_state              object
dtype: object

In [5]:
customers_dataset.isnull().sum()
# Ensures that none of the columns have no empty cells.

customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

In [6]:
print(customers_dataset.duplicated().sum())
# Ensures that there are no duplicated rows of data/cells of data.

0


In [7]:
customers_dataset['customer_city'].value_counts()

customer_city
sao paulo            15540
rio de janeiro        6882
belo horizonte        2773
brasilia              2131
curitiba              1521
                     ...  
bequimao                 1
andarai                  1
vargem grande            1
curvelandia              1
eugenio de castro        1
Name: count, Length: 4119, dtype: int64

In [8]:
customers_dataset['customer_state'].value_counts()

customer_state
SP    41746
RJ    12852
MG    11635
RS     5466
PR     5045
SC     3637
BA     3380
DF     2140
ES     2033
GO     2020
PE     1652
CE     1336
PA      975
MT      907
MA      747
MS      715
PB      536
PI      495
RN      485
AL      413
SE      350
TO      280
RO      253
AM      148
AC       81
AP       68
RR       46
Name: count, dtype: int64

### 1.2 Inspect geolocation_dataset

In [9]:
geolocation_dataset

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                               1037       -23.545621       -46.639292   
1                               1046       -23.546081       -46.644820   
2                               1046       -23.546129       -46.642951   
3                               1041       -23.544392       -46.639499   
4                               1035       -23.541578       -46.641607   
...                              ...              ...              ...   
1000158                        99950       -28.068639       -52.010705   
1000159                        99900       -27.877125       -52.224882   
1000160                        99950       -28.071855       -52.014716   
1000161                        99980       -28.388932       -51.846871   
1000162                        99950       -28.070104       -52.018658   

        geolocation_city geolocation_state  
0              sao paulo                SP  
1              sao paulo                SP  
2              sao paulo                SP  
3              sao paulo                SP  
4              sao paulo                SP  
...                  ...               ...  
1000158         tapejara                RS  
1000159   getulio vargas                RS  
1000160         tapejara                RS  
1000161  david canabarro                RS  
1000162         tapejara                RS  

[1000163 rows x 5 columns]

In [10]:
geolocation_dataset.dtypes

geolocation_zip_code_prefix      int64
geolocation_lat                float64
geolocation_lng                float64
geolocation_city                object
geolocation_state               object
dtype: object

In [11]:
geolocation_dataset.isnull().sum()
# Ensures that there are no empty cells in geolocation dataset.

geolocation_zip_code_prefix    0
geolocation_lat                0
geolocation_lng                0
geolocation_city               0
geolocation_state              0
dtype: int64

In [12]:
duplicates = geolocation_dataset[geolocation_dataset.duplicated()]
duplicates

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
15                              1046       -23.546081       -46.644820   
44                              1046       -23.546081       -46.644820   
65                              1046       -23.546081       -46.644820   
66                              1009       -23.546935       -46.636588   
67                              1046       -23.546081       -46.644820   
...                              ...              ...              ...   
1000153                        99970       -28.343273       -51.873734   
1000154                        99950       -28.070493       -52.011342   
1000159                        99900       -27.877125       -52.224882   
1000160                        99950       -28.071855       -52.014716   
1000162                        99950       -28.070104       -52.018658   

        geolocation_city geolocation_state  
15             sao paulo                SP  
44             sao paulo                SP  
65             sao paulo                SP  
66             sao paulo                SP  
67             sao paulo                SP  
...                  ...               ...  
1000153          ciriaco                RS  
1000154         tapejara                RS  
1000159   getulio vargas                RS  
1000160         tapejara                RS  
1000162         tapejara                RS  

[261831 rows x 5 columns]

### 1.3 Inspect order_items_dataset

In [13]:
order_items_dataset

order_id  order_item_id  \
0       00010242fe8c5a6d1ba2dd792cb16214              1   
1       00018f77f2f0320c557190d7a144bdd3              1   
2       000229ec398224ef6ca0657da4fc703e              1   
3       00024acbcdf0a6daa1e931b038114c75              1   
4       00042b26cf59d7ce69dfabb4e55b4fd9              1   
...                                  ...            ...   
112645  fffc94f6ce00a00581880bf54a75a037              1   
112646  fffcd46ef2263f404302a634eb57f7eb              1   
112647  fffce4705a9662cd70adb13d4a31832d              1   
112648  fffe18544ffabc95dfada21779c9644f              1   
112649  fffe41c64501cc87c801fd61db3f6244              1   

                              product_id                         seller_id  \
0       4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1       e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2       c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3       7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4       ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   
...                                  ...                               ...   
112645  4aa6014eceb682077f9dc4bffebc05b0  b8bc237ba3788b23da09c0f1f3a3288c   
112646  32e07fd915822b0765e448c4dd74c828  f3c38ab652836d21de61fb8314b69182   
112647  72a30483855e2eafc67aee5dc2560482  c3cfdc648177fdbbbb35635a37472c53   
112648  9c422a519119dcad7575db5af1ba540e  2b3e4a2a3ea8e01938cabda2a3e5cc79   
112649  350688d9dc1e75ff97be326363655e01  f7ccf836d21b2fb1de37564105216cc1   

        shipping_limit_date   price  freight_value  
0       2017-09-19 09:45:35   58.90          13.29  
1       2017-05-03 11:05:13  239.90          19.93  
2       2018-01-18 14:48:30  199.00          17.87  
3       2018-08-15 10:10:18   12.99          12.79  
4       2017-02-13 13:57:51  199.90          18.14  
...                     ...     ...            ...  
112645  2018-05-02 04:11:01  299.99          43.41  
112646  2018-07-20 04:31:48  350.00          36.53  
112647  2017-10-30 17:14:25   99.90          16.95  
112648  2017-08-21 00:04:32   55.99           8.72  
112649  2018-06-12 17:10:13   43.00          12.79  

[112650 rows x 7 columns]

In [14]:
order_items_dataset.dtypes

order_id                object
order_item_id            int64
product_id              object
seller_id               object
shipping_limit_date     object
price                  float64
freight_value          float64
dtype: object

In [15]:
order_items_dataset.isnull().sum()

order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

In [16]:
print(order_items_dataset.duplicated().sum())
print(order_items_dataset[order_items_dataset.duplicated(subset= ['order_id', 'order_item_id', 'product_id', 'seller_id'])])

0
Empty DataFrame
Columns: [order_id, order_item_id, product_id, seller_id, shipping_limit_date, price, freight_value]
Index: []


### 1.4 Inspect order_payments_dataset

In [17]:
order_payments_dataset

order_id  payment_sequential payment_type  \
0       b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1       a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2       25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3       ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4       42fdf880ba16b47b59251dd489d4441a                   1  credit_card   
...                                  ...                 ...          ...   
103881  0406037ad97740d563a178ecc7a2075c                   1       boleto   
103882  7b905861d7c825891d6347454ea7863f                   1  credit_card   
103883  32609bbb3dd69b3c066a6860554a77bf                   1  credit_card   
103884  b8b61059626efa996a60be9bb9320e10                   1  credit_card   
103885  28bbae6599b09d39ca406b747b6632b1                   1       boleto   

        payment_installments  payment_value  
0                          8          99.33  
1                          1          24.39  
2                          1          65.71  
3                          8         107.78  
4                          2         128.45  
...                      ...            ...  
103881                     1         363.31  
103882                     2          96.80  
103883                     1          47.77  
103884                     5         369.54  
103885                     1         191.58  

[103886 rows x 5 columns]

In [18]:
order_payments_dataset.dtypes

order_id                 object
payment_sequential        int64
payment_type             object
payment_installments      int64
payment_value           float64
dtype: object

In [19]:
order_payments_dataset.isnull().sum()

order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

In [20]:
print(order_payments_dataset.duplicated().sum())

0


### 1.5 Inspect order_reviews_dataset

In [21]:
order_reviews_dataset

review_id                          order_id  \
0      7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1      80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2      228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3      e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4      f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   
...                                 ...                               ...   
99219  574ed12dd733e5fa530cfd4bbf39d7c9  2a8c23fee101d4d5662fa670396eb8da   
99220  f3897127253a9592a73be9bdfdf4ed7a  22ec9f0669f784db00fa86d035cf8602   
99221  b3de70c89b1510c4cd3d0649fd302472  55d4004744368f5571d1f590031933e4   
99222  1adeb9d84d72fe4e337617733eb85149  7725825d039fc1f0ceb7635e3f7d9206   
99223  efe49f1d6f951dd88b51e6ccd4cc548f  90531360ecb1eec2a1fbb265a0db0508   

       review_score review_comment_title  \
0                 4                  NaN   
1                 5                  NaN   
2                 5                  NaN   
3                 5                  NaN   
4                 5                  NaN   
...             ...                  ...   
99219             5                  NaN   
99220             5                  NaN   
99221             5                  NaN   
99222             4                  NaN   
99223             1                  NaN   

                                  review_comment_message review_creation_date  \
0                                                    NaN  2018-01-18 00:00:00   
1                                                    NaN  2018-03-10 00:00:00   
2                                                    NaN  2018-02-17 00:00:00   
3                  Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4      Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   
...                                                  ...                  ...   
99219                                                NaN  2018-07-07 00:00:00   
99220                                                NaN  2017-12-09 00:00:00   
99221  Excelente mochila, entrega super rápida. Super...  2018-03-22 00:00:00   
99222                                                NaN  2018-07-01 00:00:00   
99223  meu produto chegou e ja tenho que devolver, po...  2017-07-03 00:00:00   

      review_answer_timestamp  
0         2018-01-18 21:46:59  
1         2018-03-11 03:05:13  
2         2018-02-18 14:36:24  
3         2017-04-21 22:02:06  
4         2018-03-02 10:26:53  
...                       ...  
99219     2018-07-14 17:18:30  
99220     2017-12-11 20:06:42  
99221     2018-03-23 09:10:43  
99222     2018-07-02 12:59:13  
99223     2017-07-03 21:01:49  

[99224 rows x 7 columns]

In [22]:
order_reviews_dataset.dtypes

review_id                  object
order_id                   object
review_score                int64
review_comment_title       object
review_comment_message     object
review_creation_date       object
review_answer_timestamp    object
dtype: object

In [23]:
order_reviews_dataset.isnull().sum()

review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_comment_message     58247
review_creation_date           0
review_answer_timestamp        0
dtype: int64

In [24]:
order_reviews_dataset[order_reviews_dataset.isnull().any(axis=1)]

review_id                          order_id  \
0      7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1      80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2      228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3      e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4      f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   
...                                 ...                               ...   
99219  574ed12dd733e5fa530cfd4bbf39d7c9  2a8c23fee101d4d5662fa670396eb8da   
99220  f3897127253a9592a73be9bdfdf4ed7a  22ec9f0669f784db00fa86d035cf8602   
99221  b3de70c89b1510c4cd3d0649fd302472  55d4004744368f5571d1f590031933e4   
99222  1adeb9d84d72fe4e337617733eb85149  7725825d039fc1f0ceb7635e3f7d9206   
99223  efe49f1d6f951dd88b51e6ccd4cc548f  90531360ecb1eec2a1fbb265a0db0508   

       review_score review_comment_title  \
0                 4                  NaN   
1                 5                  NaN   
2                 5                  NaN   
3                 5                  NaN   
4                 5                  NaN   
...             ...                  ...   
99219             5                  NaN   
99220             5                  NaN   
99221             5                  NaN   
99222             4                  NaN   
99223             1                  NaN   

                                  review_comment_message review_creation_date  \
0                                                    NaN  2018-01-18 00:00:00   
1                                                    NaN  2018-03-10 00:00:00   
2                                                    NaN  2018-02-17 00:00:00   
3                  Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4      Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   
...                                                  ...                  ...   
99219                                                NaN  2018-07-07 00:00:00   
99220                                                NaN  2017-12-09 00:00:00   
99221  Excelente mochila, entrega super rápida. Super...  2018-03-22 00:00:00   
99222                                                NaN  2018-07-01 00:00:00   
99223  meu produto chegou e ja tenho que devolver, po...  2017-07-03 00:00:00   

      review_answer_timestamp  
0         2018-01-18 21:46:59  
1         2018-03-11 03:05:13  
2         2018-02-18 14:36:24  
3         2017-04-21 22:02:06  
4         2018-03-02 10:26:53  
...                       ...  
99219     2018-07-14 17:18:30  
99220     2017-12-11 20:06:42  
99221     2018-03-23 09:10:43  
99222     2018-07-02 12:59:13  
99223     2017-07-03 21:01:49  

[89385 rows x 7 columns]

In [25]:
order_reviews_dataset.dropna()
order_reviews_dataset.reset_index()
order_reviews_dataset

review_id                          order_id  \
0      7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1      80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2      228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3      e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4      f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   
...                                 ...                               ...   
99219  574ed12dd733e5fa530cfd4bbf39d7c9  2a8c23fee101d4d5662fa670396eb8da   
99220  f3897127253a9592a73be9bdfdf4ed7a  22ec9f0669f784db00fa86d035cf8602   
99221  b3de70c89b1510c4cd3d0649fd302472  55d4004744368f5571d1f590031933e4   
99222  1adeb9d84d72fe4e337617733eb85149  7725825d039fc1f0ceb7635e3f7d9206   
99223  efe49f1d6f951dd88b51e6ccd4cc548f  90531360ecb1eec2a1fbb265a0db0508   

       review_score review_comment_title  \
0                 4                  NaN   
1                 5                  NaN   
2                 5                  NaN   
3                 5                  NaN   
4                 5                  NaN   
...             ...                  ...   
99219             5                  NaN   
99220             5                  NaN   
99221             5                  NaN   
99222             4                  NaN   
99223             1                  NaN   

                                  review_comment_message review_creation_date  \
0                                                    NaN  2018-01-18 00:00:00   
1                                                    NaN  2018-03-10 00:00:00   
2                                                    NaN  2018-02-17 00:00:00   
3                  Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4      Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   
...                                                  ...                  ...   
99219                                                NaN  2018-07-07 00:00:00   
99220                                                NaN  2017-12-09 00:00:00   
99221  Excelente mochila, entrega super rápida. Super...  2018-03-22 00:00:00   
99222                                                NaN  2018-07-01 00:00:00   
99223  meu produto chegou e ja tenho que devolver, po...  2017-07-03 00:00:00   

      review_answer_timestamp  
0         2018-01-18 21:46:59  
1         2018-03-11 03:05:13  
2         2018-02-18 14:36:24  
3         2017-04-21 22:02:06  
4         2018-03-02 10:26:53  
...                       ...  
99219     2018-07-14 17:18:30  
99220     2017-12-11 20:06:42  
99221     2018-03-23 09:10:43  
99222     2018-07-02 12:59:13  
99223     2017-07-03 21:01:49  

[99224 rows x 7 columns]

In [26]:
order_reviews_dataset.duplicated().sum()

np.int64(0)

### 1.6 Inspect orders_dataset

In [27]:
orders_dataset

order_id                       customer_id  \
0      e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1      53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2      47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3      949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4      ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                 ...                               ...   
99436  9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
99437  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
99438  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
99439  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
99440  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

      order_status order_purchase_timestamp    order_approved_at  \
0        delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1        delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2        delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3        delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4        delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   
...            ...                      ...                  ...   
99436    delivered      2017-03-09 09:54:05  2017-03-09 09:54:05   
99437    delivered      2018-02-06 12:58:58  2018-02-06 13:10:37   
99438    delivered      2017-08-27 14:46:43  2017-08-27 15:04:16   
99439    delivered      2018-01-08 21:28:27  2018-01-08 21:36:21   
99440    delivered      2018-03-08 20:57:30  2018-03-09 11:20:28   

      order_delivered_carrier_date order_delivered_customer_date  \
0              2017-10-04 19:55:00           2017-10-10 21:25:13   
1              2018-07-26 14:31:00           2018-08-07 15:27:45   
2              2018-08-08 13:50:00           2018-08-17 18:06:29   
3              2017-11-22 13:39:59           2017-12-02 00:28:42   
4              2018-02-14 19:46:34           2018-02-16 18:17:02   
...                            ...                           ...   
99436          2017-03-10 11:18:03           2017-03-17 15:08:01   
99437          2018-02-07 23:22:42           2018-02-28 17:37:56   
99438          2017-08-28 20:52:26           2017-09-21 11:24:17   
99439          2018-01-12 15:35:03           2018-01-25 23:32:54   
99440          2018-03-09 22:11:59           2018-03-16 13:08:30   

      order_estimated_delivery_date  
0               2017-10-18 00:00:00  
1               2018-08-13 00:00:00  
2               2018-09-04 00:00:00  
3               2017-12-15 00:00:00  
4               2018-02-26 00:00:00  
...                             ...  
99436           2017-03-28 00:00:00  
99437           2018-03-02 00:00:00  
99438           2017-09-27 00:00:00  
99439           2018-02-15 00:00:00  
99440           2018-04-03 00:00:00  

[99441 rows x 8 columns]

In [28]:
orders_dataset.dtypes

order_id                         object
customer_id                      object
order_status                     object
order_purchase_timestamp         object
order_approved_at                object
order_delivered_carrier_date     object
order_delivered_customer_date    object
order_estimated_delivery_date    object
dtype: object

In [29]:
orders_dataset.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [30]:
orders_dataset[orders_dataset.isnull().any(axis=1)]

order_id                       customer_id  \
6      136cce7faa42fdb2cefd53fdc79a6098  ed0271e0b7da060a393796590e7b737a   
44     ee64d42b8cf066f35eac1cf57de1aa85  caded193e8e47b8362864762a83db3c5   
103    0760a852e4e9d89eb77bf631eaaf1c84  d2a79636084590b7465af8ab374a8cf5   
128    15bed8e2fec7fdbadb186b57c46c92f2  f3f0e613e0bdb9c7cee75504f0f90679   
154    6942b8da583c2f9957e990d028607019  52006a9383bf149a4fb24226b173106f   
...                                 ...                               ...   
99283  3a3cddda5a7c27851bd96c3313412840  0b0d6095c5555fe083844281f6b093bb   
99313  e9e64a17afa9653aacf2616d94c005b8  b4cd0522e632e481f8eaf766a2646e86   
99347  a89abace0dcc01eeb267a9660b5ac126  2f0524a7b1b3845a1a57fcf3910c4333   
99348  a69ba794cc7deb415c3e15a0a3877e69  726f0894b5becdf952ea537d5266e543   
99415  5fabc81b6322c8443648e1b21a6fef21  32c9df889d41b0ee8309a5efb6855dcb   

      order_status order_purchase_timestamp    order_approved_at  \
6         invoiced      2017-04-11 12:22:08  2017-04-13 13:25:17   
44         shipped      2018-06-04 16:44:48  2018-06-05 04:31:18   
103       invoiced      2018-08-03 17:44:42  2018-08-07 06:15:14   
128     processing      2017-09-03 14:22:03  2017-09-03 14:30:09   
154        shipped      2018-01-10 11:33:07  2018-01-11 02:32:30   
...            ...                      ...                  ...   
99283     canceled      2018-08-31 16:13:44                  NaN   
99313   processing      2018-01-05 23:07:24  2018-01-09 07:18:05   
99347     canceled      2018-09-06 18:45:47                  NaN   
99348  unavailable      2017-08-23 16:28:04  2017-08-28 15:44:47   
99415  unavailable      2017-10-10 10:50:03  2017-10-14 18:35:57   

      order_delivered_carrier_date order_delivered_customer_date  \
6                              NaN                           NaN   
44             2018-06-05 14:32:00                           NaN   
103                            NaN                           NaN   
128                            NaN                           NaN   
154            2018-01-11 19:39:23                           NaN   
...                            ...                           ...   
99283                          NaN                           NaN   
99313                          NaN                           NaN   
99347                          NaN                           NaN   
99348                          NaN                           NaN   
99415                          NaN                           NaN   

      order_estimated_delivery_date  
6               2017-05-09 00:00:00  
44              2018-06-28 00:00:00  
103             2018-08-21 00:00:00  
128             2017-10-03 00:00:00  
154             2018-02-07 00:00:00  
...                             ...  
99283           2018-10-01 00:00:00  
99313           2018-02-06 00:00:00  
99347           2018-09-27 00:00:00  
99348           2017-09-15 00:00:00  
99415           2017-10-23 00:00:00  

[2980 rows x 8 columns]

In [31]:
orders_dataset = orders_dataset.dropna()
orders_dataset.reset_index()
orders_dataset

order_id                       customer_id  \
0      e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1      53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2      47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3      949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4      ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                 ...                               ...   
99436  9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
99437  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
99438  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
99439  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
99440  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

      order_status order_purchase_timestamp    order_approved_at  \
0        delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1        delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2        delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3        delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4        delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   
...            ...                      ...                  ...   
99436    delivered      2017-03-09 09:54:05  2017-03-09 09:54:05   
99437    delivered      2018-02-06 12:58:58  2018-02-06 13:10:37   
99438    delivered      2017-08-27 14:46:43  2017-08-27 15:04:16   
99439    delivered      2018-01-08 21:28:27  2018-01-08 21:36:21   
99440    delivered      2018-03-08 20:57:30  2018-03-09 11:20:28   

      order_delivered_carrier_date order_delivered_customer_date  \
0              2017-10-04 19:55:00           2017-10-10 21:25:13   
1              2018-07-26 14:31:00           2018-08-07 15:27:45   
2              2018-08-08 13:50:00           2018-08-17 18:06:29   
3              2017-11-22 13:39:59           2017-12-02 00:28:42   
4              2018-02-14 19:46:34           2018-02-16 18:17:02   
...                            ...                           ...   
99436          2017-03-10 11:18:03           2017-03-17 15:08:01   
99437          2018-02-07 23:22:42           2018-02-28 17:37:56   
99438          2017-08-28 20:52:26           2017-09-21 11:24:17   
99439          2018-01-12 15:35:03           2018-01-25 23:32:54   
99440          2018-03-09 22:11:59           2018-03-16 13:08:30   

      order_estimated_delivery_date  
0               2017-10-18 00:00:00  
1               2018-08-13 00:00:00  
2               2018-09-04 00:00:00  
3               2017-12-15 00:00:00  
4               2018-02-26 00:00:00  
...                             ...  
99436           2017-03-28 00:00:00  
99437           2018-03-02 00:00:00  
99438           2017-09-27 00:00:00  
99439           2018-02-15 00:00:00  
99440           2018-04-03 00:00:00  

[96461 rows x 8 columns]

In [32]:
print(orders_dataset.duplicated().sum())

0


### 1.7 Inspect products_dataset

In [33]:
products_dataset

product_id              product_category_name  \
0      1e9e8ef04dbcff4541ed26657ea517e5                         perfumaria   
1      3aa071139cb16b67ca9e5dea641aaa2f                              artes   
2      96bd76ec8810374ed1b65e291975717f                      esporte_lazer   
3      cef67bcfe19066a932b7673e239eb23d                              bebes   
4      9dc1a7de274444849c219cff195d0b71              utilidades_domesticas   
...                                 ...                                ...   
32946  a0b7d5a992ccda646f2d34e418fff5a0                   moveis_decoracao   
32947  bf4538d88321d0fd4412a93c974510e6  construcao_ferramentas_iluminacao   
32948  9a7c6041fa9592d9d9ef6cfe62a71f8c                    cama_mesa_banho   
32949  83808703fc0706a22e264b9d75f04a2e             informatica_acessorios   
32950  106392145fca363410d287a815be6de4                    cama_mesa_banho   

       product_name_lenght  product_description_lenght  product_photos_qty  \
0                     40.0                       287.0                 1.0   
1                     44.0                       276.0                 1.0   
2                     46.0                       250.0                 1.0   
3                     27.0                       261.0                 1.0   
4                     37.0                       402.0                 4.0   
...                    ...                         ...                 ...   
32946                 45.0                        67.0                 2.0   
32947                 41.0                       971.0                 1.0   
32948                 50.0                       799.0                 1.0   
32949                 60.0                       156.0                 2.0   
32950                 58.0                       309.0                 1.0   

       product_weight_g  product_length_cm  product_height_cm  \
0                 225.0               16.0               10.0   
1                1000.0               30.0               18.0   
2                 154.0               18.0                9.0   
3                 371.0               26.0                4.0   
4                 625.0               20.0               17.0   
...                 ...                ...                ...   
32946           12300.0               40.0               40.0   
32947            1700.0               16.0               19.0   
32948            1400.0               27.0                7.0   
32949             700.0               31.0               13.0   
32950            2083.0               12.0                2.0   

       product_width_cm  
0                  14.0  
1                  20.0  
2                  15.0  
3                  26.0  
4                  13.0  
...                 ...  
32946              40.0  
32947              16.0  
32948              27.0  
32949              20.0  
32950               7.0  

[32951 rows x 9 columns]

In [34]:
products_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB


In [35]:
products_dataset.isnull().sum()

product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

In [36]:
products_dataset[products_dataset.isnull().any(axis=1)]

product_id product_category_name  \
105    a41e356c76fab66334f36de622ecbd3a                   NaN   
128    d8dee61c2034d6d075997acef1870e9b                   NaN   
145    56139431d72cd51f19eb9f7dae4d1617                   NaN   
154    46b48281eb6d663ced748f324108c733                   NaN   
197    5fb61f482620cb672f5e586bb132eae9                   NaN   
...                                 ...                   ...   
32515  b0a0c5dd78e644373b199380612c350a                   NaN   
32589  10dbe0fbaa2c505123c17fdc34a63c56                   NaN   
32616  bd2ada37b58ae94cc838b9c0569fecd8                   NaN   
32772  fa51e914046aab32764c41356b9d4ea4                   NaN   
32852  c4ceee876c82b8328e9c293fa0e1989b                   NaN   

       product_name_lenght  product_description_lenght  product_photos_qty  \
105                    NaN                         NaN                 NaN   
128                    NaN                         NaN                 NaN   
145                    NaN                         NaN                 NaN   
154                    NaN                         NaN                 NaN   
197                    NaN                         NaN                 NaN   
...                    ...                         ...                 ...   
32515                  NaN                         NaN                 NaN   
32589                  NaN                         NaN                 NaN   
32616                  NaN                         NaN                 NaN   
32772                  NaN                         NaN                 NaN   
32852                  NaN                         NaN                 NaN   

       product_weight_g  product_length_cm  product_height_cm  \
105               650.0               17.0               14.0   
128               300.0               16.0                7.0   
145               200.0               20.0               20.0   
154             18500.0               41.0               30.0   
197               300.0               35.0                7.0   
...                 ...                ...                ...   
32515            1800.0               30.0               20.0   
32589             800.0               30.0               10.0   
32616             200.0               21.0                8.0   
32772            1300.0               45.0               16.0   
32852             700.0               28.0                3.0   

       product_width_cm  
105                12.0  
128                20.0  
145                20.0  
154                41.0  
197                12.0  
...                 ...  
32515              70.0  
32589              23.0  
32616              16.0  
32772              45.0  
32852              43.0  

[611 rows x 9 columns]

In [37]:
# Rows missing from group A (the 610 nows that have null values)
group_a = products_dataset[products_dataset[['product_category_name', 'product_name_lenght',
                 'product_description_lenght', 'product_photos_qty']].isnull().any(axis=1)]

# Rows missing from group B (the 2 nows that have null values)
group_b = products_dataset[products_dataset[['product_weight_g', 'product_length_cm',
                 'product_height_cm', 'product_width_cm']].isnull().any(axis=1)]

# Intersection of A and B
intersection = pd.merge(group_a, group_b, how='inner')

print(f"Overlap between the two groups: {len(intersection)} row(s)")


Overlap between the two groups: 1 row(s)


In [38]:
print(products_dataset.duplicated().sum())

0


Understand that products_dataset has:
- 32,951 rows and 9 columns.
- There are 610 rows that have missing values in columns: product_category_name, product_name_lenght, product_description_lenght, product_photos_qty. (likely belong to the same group of products)
- There are 2 missing values in product_weight_g, product_length_cm, product_height_cm, product_width_cm.(likely belong to the same group of products)
- There is 1 overlapping row between the two groups, meaning there is 1 row that appears in both columns, assuming the grouped columns are always null together.
- No duplicated rows.
- Possible typos of 2 columns: product_name_lenght and product_description_lenght.

### 1.8 Inspect sellers_dataset

In [39]:
sellers_dataset

seller_id  seller_zip_code_prefix  \
0     3442f8959a84dea7ee197c632cb2df15                   13023   
1     d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
2     ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
3     c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
4     51a04a8a6bdcb23deccc82b0b80742cf                   12914   
...                                ...                     ...   
3090  98dddbc4601dd4443ca174359b237166                   87111   
3091  f8201cab383e484733266d1906e2fdfa                   88137   
3092  74871d19219c7d518d0090283e03c137                    4650   
3093  e603cf3fec55f8697c9059638d6c8eb5                   96080   
3094  9e25199f6ef7e7c347120ff175652c3b                   12051   

            seller_city seller_state  
0              campinas           SP  
1            mogi guacu           SP  
2        rio de janeiro           RJ  
3             sao paulo           SP  
4     braganca paulista           SP  
...                 ...          ...  
3090            sarandi           PR  
3091            palhoca           SC  
3092          sao paulo           SP  
3093            pelotas           RS  
3094            taubate           SP  

[3095 rows x 4 columns]

In [40]:
sellers_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   seller_id               3095 non-null   object
 1   seller_zip_code_prefix  3095 non-null   int64 
 2   seller_city             3095 non-null   object
 3   seller_state            3095 non-null   object
dtypes: int64(1), object(3)
memory usage: 96.8+ KB


In [41]:
sellers_dataset.isnull().sum()

seller_id                 0
seller_zip_code_prefix    0
seller_city               0
seller_state              0
dtype: int64

In [42]:
print(sellers_dataset.duplicated().sum())

0


Understand that sellers_dataset has:
- 3095 rows and 4 columns.
- No missing values.
- No duplicated rows.

### 1.9 Inspect product_category_name_translation_dataset

In [43]:
product_category_name_translation_dataset

product_category_name product_category_name_english
0                    beleza_saude                 health_beauty
1          informatica_acessorios         computers_accessories
2                      automotivo                          auto
3                 cama_mesa_banho                bed_bath_table
4                moveis_decoracao               furniture_decor
..                            ...                           ...
66                         flores                       flowers
67             artes_e_artesanato         arts_and_craftmanship
68                fraldas_higiene           diapers_and_hygiene
69  fashion_roupa_infanto_juvenil     fashion_childrens_clothes
70             seguros_e_servicos         security_and_services

[71 rows x 2 columns]

In [44]:
product_category_name_translation_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   product_category_name          71 non-null     object
 1   product_category_name_english  71 non-null     object
dtypes: object(2)
memory usage: 1.2+ KB


In [45]:
product_category_name_translation_dataset.isnull().sum()

product_category_name            0
product_category_name_english    0
dtype: int64

In [46]:
print(product_category_name_translation_dataset.duplicated().sum())

0


Understand that product_category_name_translation_dataset has:
- 71 rows and 2 columns.
- No missing values.
- No duplicated rows.

## 2. Data Cleaning

### 2.1 Clean geolocation_dataset (1.2)

In [47]:
geolocation_dataset = geolocation_dataset.drop_duplicates()
geolocation_dataset

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                               1037       -23.545621       -46.639292   
1                               1046       -23.546081       -46.644820   
2                               1046       -23.546129       -46.642951   
3                               1041       -23.544392       -46.639499   
4                               1035       -23.541578       -46.641607   
...                              ...              ...              ...   
1000155                        99965       -28.180655       -52.034367   
1000156                        99950       -28.072188       -52.011272   
1000157                        99950       -28.068864       -52.012964   
1000158                        99950       -28.068639       -52.010705   
1000161                        99980       -28.388932       -51.846871   

        geolocation_city geolocation_state  
0              sao paulo                SP  
1              sao paulo                SP  
2              sao paulo                SP  
3              sao paulo                SP  
4              sao paulo                SP  
...                  ...               ...  
1000155       agua santa                RS  
1000156         tapejara                RS  
1000157         tapejara                RS  
1000158         tapejara                RS  
1000161  david canabarro                RS  

[738332 rows x 5 columns]

print(geolocation_dataset[geolocation_dataset.duplicated(subset=['geolocation_lat', 'geolocation_lng'])])

In [48]:
print(geolocation_dataset[geolocation_dataset.duplicated(subset=['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng'])])
geolocation_dataset = geolocation_dataset.drop_duplicates(subset=['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng'])
geolocation_dataset = geolocation_dataset.reset_index(drop=True)
geolocation_dataset

         geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
154                             1043       -23.544735       -46.642523   
159                             1050       -23.548802       -46.642373   
162                             1024       -23.541390       -46.629899   
181                             1042       -23.544667       -46.640669   
192                             1020       -23.552496       -46.632060   
...                              ...              ...              ...   
1000046                        99955       -28.107588       -52.144019   
1000053                        99900       -27.888360       -52.224029   
1000094                        99940       -28.060955       -51.858637   
1000147                        99900       -27.880973       -52.223206   
1000149                        99900       -27.882555       -52.222879   

        geolocation_city geolocation_state  
154            são paulo                SP  
159            sao pa

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                              1037       -23.545621       -46.639292   
1                              1046       -23.546081       -46.644820   
2                              1046       -23.546129       -46.642951   
3                              1041       -23.544392       -46.639499   
4                              1035       -23.541578       -46.641607   
...                             ...              ...              ...   
720149                        99965       -28.180655       -52.034367   
720150                        99950       -28.072188       -52.011272   
720151                        99950       -28.068864       -52.012964   
720152                        99950       -28.068639       -52.010705   
720153                        99980       -28.388932       -51.846871   

       geolocation_city geolocation_state  
0             sao paulo                SP  
1             sao paulo                SP  
2             sao paulo                SP  
3             sao paulo                SP  
4             sao paulo                SP  
...                 ...               ...  
720149       agua santa                RS  
720150         tapejara                RS  
720151         tapejara                RS  
720152         tapejara                RS  
720153  david canabarro                RS  

[720154 rows x 5 columns]

In [49]:
geolocation_dataset.drop(columns=['geolocation_lat', 'geolocation_lng'])

geolocation_zip_code_prefix geolocation_city geolocation_state
0                              1037        sao paulo                SP
1                              1046        sao paulo                SP
2                              1046        sao paulo                SP
3                              1041        sao paulo                SP
4                              1035        sao paulo                SP
...                             ...              ...               ...
720149                        99965       agua santa                RS
720150                        99950         tapejara                RS
720151                        99950         tapejara                RS
720152                        99950         tapejara                RS
720153                        99980  david canabarro                RS

[720154 rows x 3 columns]

### 2.2 Clean order_reviews_dataset (1.5)

In [50]:
order_reviews_dataset.drop(columns=['review_comment_title', 'review_comment_message'])

review_id                          order_id  \
0      7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1      80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2      228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3      e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4      f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   
...                                 ...                               ...   
99219  574ed12dd733e5fa530cfd4bbf39d7c9  2a8c23fee101d4d5662fa670396eb8da   
99220  f3897127253a9592a73be9bdfdf4ed7a  22ec9f0669f784db00fa86d035cf8602   
99221  b3de70c89b1510c4cd3d0649fd302472  55d4004744368f5571d1f590031933e4   
99222  1adeb9d84d72fe4e337617733eb85149  7725825d039fc1f0ceb7635e3f7d9206   
99223  efe49f1d6f951dd88b51e6ccd4cc548f  90531360ecb1eec2a1fbb265a0db0508   

       review_score review_creation_date review_answer_timestamp  
0                 4  2018-01-18 00:00:00     2018-01-18 21:46:59  
1                 5  2018-03-10 00:00:00     2018-03-11 03:05:13  
2                 5  2018-02-17 00:00:00     2018-02-18 14:36:24  
3                 5  2017-04-21 00:00:00     2017-04-21 22:02:06  
4                 5  2018-03-01 00:00:00     2018-03-02 10:26:53  
...             ...                  ...                     ...  
99219             5  2018-07-07 00:00:00     2018-07-14 17:18:30  
99220             5  2017-12-09 00:00:00     2017-12-11 20:06:42  
99221             5  2018-03-22 00:00:00     2018-03-23 09:10:43  
99222             4  2018-07-01 00:00:00     2018-07-02 12:59:13  
99223             1  2017-07-03 00:00:00     2017-07-03 21:01:49  

[99224 rows x 5 columns]

### 2.3 Clean products_dataset (1.7)

In [51]:
# 1.7 Drop columns product_name_lenght and product_description_lenght. Decided to drop because it is redundant to the project objective of identifying potential repeat buyers.

products_dataset.drop(columns=['product_name_lenght', 'product_description_lenght'])

product_id              product_category_name  \
0      1e9e8ef04dbcff4541ed26657ea517e5                         perfumaria   
1      3aa071139cb16b67ca9e5dea641aaa2f                              artes   
2      96bd76ec8810374ed1b65e291975717f                      esporte_lazer   
3      cef67bcfe19066a932b7673e239eb23d                              bebes   
4      9dc1a7de274444849c219cff195d0b71              utilidades_domesticas   
...                                 ...                                ...   
32946  a0b7d5a992ccda646f2d34e418fff5a0                   moveis_decoracao   
32947  bf4538d88321d0fd4412a93c974510e6  construcao_ferramentas_iluminacao   
32948  9a7c6041fa9592d9d9ef6cfe62a71f8c                    cama_mesa_banho   
32949  83808703fc0706a22e264b9d75f04a2e             informatica_acessorios   
32950  106392145fca363410d287a815be6de4                    cama_mesa_banho   

       product_photos_qty  product_weight_g  product_length_cm  \
0                     1.0             225.0               16.0   
1                     1.0            1000.0               30.0   
2                     1.0             154.0               18.0   
3                     1.0             371.0               26.0   
4                     4.0             625.0               20.0   
...                   ...               ...                ...   
32946                 2.0           12300.0               40.0   
32947                 1.0            1700.0               16.0   
32948                 1.0            1400.0               27.0   
32949                 2.0             700.0               31.0   
32950                 1.0            2083.0               12.0   

       product_height_cm  product_width_cm  
0                   10.0              14.0  
1                   18.0              20.0  
2                    9.0              15.0  
3                    4.0              26.0  
4                   17.0              13.0  
...                  ...               ...  
32946               40.0              40.0  
32947               19.0              16.0  
32948                7.0              27.0  
32949               13.0              20.0  
32950                2.0               7.0  

[32951 rows x 7 columns]

## 3. Dataset Merging: Build a Master Customer Table

Merge pathway:

orders_dataset as the main transaction table. 

Join: 
1) customers_dataset on customer_id
2) geolocation_dataset on customer_zip_code_prefix
3) order_items_dataset on order_id
4) order_payments_dataset on order_id
5) order_reviews_dataset on order_id
6) products_dataset on product_id
7) sellers_dataset on geolocation_zip_code_prefix
8) product_category_name_translation_dataset on product_category_name 

In [52]:
pd.DataFrame({'x': [1, 2, 3], 'name': ['Alice', 'Bob', 'Charlie']})
order_items_dataset,customers_dataset.columns, geolocation_dataset.columns, order_items_dataset.columns, order_payments_dataset.columns, order_reviews_dataset.columns, orders_dataset.columns, products_dataset.columns, sellers_dataset.columns, product_category_name_translation_dataset.columns        

(                                order_id  order_item_id  \
 0       00010242fe8c5a6d1ba2dd792cb16214              1   
 1       00018f77f2f0320c557190d7a144bdd3              1   
 2       000229ec398224ef6ca0657da4fc703e              1   
 3       00024acbcdf0a6daa1e931b038114c75              1   
 4       00042b26cf59d7ce69dfabb4e55b4fd9              1   
 ...                                  ...            ...   
 112645  fffc94f6ce00a00581880bf54a75a037              1   
 112646  fffcd46ef2263f404302a634eb57f7eb              1   
 112647  fffce4705a9662cd70adb13d4a31832d              1   
 112648  fffe18544ffabc95dfada21779c9644f              1   
 112649  fffe41c64501cc87c801fd61db3f6244              1   
 
                               product_id                         seller_id  \
 0       4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
 1       e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
 2       c777355d18b72b67abbeef9df44fd0fd

In [66]:
# customers_dataset['customer_zip_code_prefix']
customers_dataset

customer_id                customer_unique_id  \
0      06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1      18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2      4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3      b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4      4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
...                                 ...                               ...   
99436  17ddf5dd5d51696bb3d7c6291687be6f  1a29b476fee25c95fbafc67c5ac95cf8   
99437  e7b71a9017aa05c9a7fd292d714858e8  d52a67c98be1cf6a5c84435bd38d095d   
99438  5e28dfe12db7fb50a4b2f691faecea5e  e9f50caf99f032f0bf3c55141f019d99   
99439  56b18e2166679b8a959d72dd06da27f9  73c2643a0a458b49f58cea58833b192e   
99440  274fa6071e5e17fe303b9748641082c8  84732c5050c01db9b23e19ba39899398   

       zip_code_prefix          customer_city customer_state  
0                14409                 franca             SP  
1                 9790  sao bernardo do campo             SP  
2                 1151              sao paulo             SP  
3                 8775        mogi das cruzes             SP  
4                13056               campinas             SP  
...                ...                    ...            ...  
99436             3937              sao paulo             SP  
99437             6764        taboao da serra             SP  
99438            60115              fortaleza             CE  
99439            92120                 canoas             RS  
99440             6703                  cotia             SP  

[99441 rows x 5 columns]

In [65]:
# geolocation_dataset['geolocation_zip_code_prefix']
geolocation_dataset

zip_code_prefix  geolocation_lat  geolocation_lng geolocation_city  \
0                  1037       -23.545621       -46.639292        sao paulo   
1                  1046       -23.546081       -46.644820        sao paulo   
2                  1046       -23.546129       -46.642951        sao paulo   
3                  1041       -23.544392       -46.639499        sao paulo   
4                  1035       -23.541578       -46.641607        sao paulo   
...                 ...              ...              ...              ...   
720149            99965       -28.180655       -52.034367       agua santa   
720150            99950       -28.072188       -52.011272         tapejara   
720151            99950       -28.068864       -52.012964         tapejara   
720152            99950       -28.068639       -52.010705         tapejara   
720153            99980       -28.388932       -51.846871  david canabarro   

       geolocation_state  
0                     SP  
1                     SP  
2                     SP  
3                     SP  
4                     SP  
...                  ...  
720149                RS  
720150                RS  
720151                RS  
720152                RS  
720153                RS  

[720154 rows x 5 columns]

In [64]:
customers_dataset.rename(columns={'customer_zip_code_prefix': 'zip_code_prefix'}, inplace=True)
geolocation_dataset.rename(columns={'geolocation_zip_code_prefix': 'zip_code_prefix'}, inplace=True)

In [71]:
sellers_dataset.rename(columns={'seller_zip_code_prefix': 'zip_code_prefix'}, inplace=True)
sellers_dataset

seller_id  zip_code_prefix        seller_city  \
0     3442f8959a84dea7ee197c632cb2df15            13023           campinas   
1     d1b65fc7debc3361ea86b5f14c68d2e2            13844         mogi guacu   
2     ce3ad9de960102d0677a81f5d0bb7b2d            20031     rio de janeiro   
3     c0f3eea2e14555b6faeea3dd58c1b1c3             4195          sao paulo   
4     51a04a8a6bdcb23deccc82b0b80742cf            12914  braganca paulista   
...                                ...              ...                ...   
3090  98dddbc4601dd4443ca174359b237166            87111            sarandi   
3091  f8201cab383e484733266d1906e2fdfa            88137            palhoca   
3092  74871d19219c7d518d0090283e03c137             4650          sao paulo   
3093  e603cf3fec55f8697c9059638d6c8eb5            96080            pelotas   
3094  9e25199f6ef7e7c347120ff175652c3b            12051            taubate   

     seller_state  
0              SP  
1              SP  
2              RJ  
3              SP  
4              SP  
...           ...  
3090           PR  
3091           SC  
3092           SP  
3093           RS  
3094           SP  

[3095 rows x 4 columns]

In [72]:
# Merge orders_dataset with customers_dataset. 
master_dataset = pd.merge(orders_dataset, customers_dataset, on='customer_id', how='left')

# Merge with geolocation_dataset.
master_dataset = pd.merge(master_dataset, geolocation_dataset, on='zip_code_prefix', how='left')

# Merge with order_items_dataset.
master_dataset = pd.merge(master_dataset, order_items_dataset, on='order_id', how='left')

# Merge with order_payments_dataset.
master_dataset = pd.merge(master_dataset, order_payments_dataset, on='order_id', how='left')

# Merge with order_reviews_dataset.
master_dataset = pd.merge(master_dataset, order_reviews_dataset, on='order_id', how='left')

# Merge with products_dataset.
master_dataset = pd.merge(master_dataset, products_dataset, on='product_id', how='left')

# Merge with sellers_dataset.
master_dataset = pd.merge(master_dataset, sellers_dataset, on='zip_code_prefix', how='left')

# Merge with product_category_name_translation_dataset.
master_dataset = pd.merge(master_dataset, product_category_name_translation_dataset, on='product_category_name', how='left')

In [74]:
master_dataset.columns

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date',
       'customer_unique_id', 'zip_code_prefix', 'customer_city',
       'customer_state', 'geolocation_lat', 'geolocation_lng',
       'geolocation_city', 'geolocation_state', 'order_item_id', 'product_id',
       'seller_id_x', 'shipping_limit_date', 'price', 'freight_value',
       'payment_sequential', 'payment_type', 'payment_installments',
       'payment_value', 'review_id', 'review_score', 'review_comment_title',
       'review_comment_message', 'review_creation_date',
       'review_answer_timestamp', 'product_category_name',
       'product_name_lenght', 'product_description_lenght',
       'product_photos_qty', 'product_weight_g', 'product_length_cm',
       'product_height_cm', 'product_width_cm', 'seller_id_y', 'seller_city',
       'seller_state', 'product_category_

In [75]:
master_dataset

order_id                       customer_id  \
0         e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1         e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2         e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3         e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
4         e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
...                                    ...                               ...   
16063068  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   
16063069  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   
16063070  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   
16063071  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   
16063072  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

         order_status order_purchase_timestamp    order_approved_at  \
0           delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1           delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
2           delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
3           delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
4           delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
...               ...                      ...                  ...   
16063068    delivered      2018-03-08 20:57:30  2018-03-09 11:20:28   
16063069    delivered      2018-03-08 20:57:30  2018-03-09 11:20:28   
16063070    delivered      2018-03-08 20:57:30  2018-03-09 11:20:28   
16063071    delivered      2018-03-08 20:57:30  2018-03-09 11:20:28   
16063072    delivered      2018-03-08 20:57:30  2018-03-09 11:20:28   

         order_delivered_carrier_date order_delivered_customer_date  \
0                 2017-10-04 19:55:00           2017-10-10 21:25:13   
1                 2017-10-04 19:55:00           2017-10-10 21:25:13   
2                 2017-10-04 19:55:00           2017-10-10 21:25:13   
3                 2017-10-04 19:55:00           2017-10-10 21:25:13   
4                 2017-10-04 19:55:00           2017-10-10 21:25:13   
...                               ...                           ...   
16063068          2018-03-09 22:11:59           2018-03-16 13:08:30   
16063069          2018-03-09 22:11:59           2018-03-16 13:08:30   
16063070          2018-03-09 22:11:59           2018-03-16 13:08:30   
16063071          2018-03-09 22:11:59           2018-03-16 13:08:30   
16063072          2018-03-09 22:11:59           2018-03-16 13:08:30   

         order_estimated_delivery_date                customer_unique_id  \
0                  2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
1                  2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
2                  2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
3                  2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
4                  2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
...                                ...                               ...   
16063068           2018-04-03 00:00:00  60350aa974b26ff12caad89e55993bd6   
16063069           2018-04-03 00:00:00  60350aa974b26ff12caad89e55993bd6   
16063070           2018-04-03 00:00:00  60350aa974b26ff12caad89e55993bd6   
16063071           2018-04-03 00:00:00  60350aa974b26ff12caad89e55993bd6   
16063072           2018-04-03 00:00:00  60350aa974b26ff12caad89e55993bd6   

          zip_code_prefix  ... product_description_lenght product_photos_qty  \
0                    3149  ...                      268.0                4.0   
1                    3149  ...                      268.0                4.0   
2                    3149  ...                      268.0                4.0   
3                    3149  ...                      268.0                4.0   
4                    3149  ...                      26

## 4. Feature Engineering

##  5. Visualization

## 6. Insights Summary